### ISSUE TITLE
Auto (de)ref not always handled correctly (type-mismatch false positive)

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/rust-analyzer#11948]
When opening the [Miri](https://github.com/rust-lang/miri) source code at revision c568f32f, and opening `src/diagnostics.rs` in vscode, an error is reported in [this line](https://github.com/rust-lang/miri/blob/6e1f3cd8ff080b85f6ba71b3f90300aca194a434/src/diagnostics.rs#L219): "expected &String, found String".

```
[{
	"resource": "/home/r/src/rust/miri/src/diagnostics.rs",
	"owner": "rustc",
	"code": {
		"value": "type-mismatch",
		"target": {
			"$mid": 1,
			"external": "https://rust-analyzer.github.io/manual.html#type-mismatch",
			"path": "/manual.html",
			"scheme": "https",
			"authority": "rust-analyzer.github.io",
			"fragment": "type-mismatch"
		}
	},
	"severity": 8,
	"message": "expected &String, found String",
	"source": "rust-analyzer",
	"startLineNumber": 219,
	"startColumn": 19,
	"endLineNumber": 219,
	"endColumn": 32
}]
```

I also saw a similar error in a file somewhere in the rustc repository but forgot where...

**rust-analyzer version**: rust-analyzer version: 46d7ee68f 2022-04-04 stable

**rustc version**: rustc 1.62.0-nightly (fbdb10f9f 2022-04-09)

**relevant settings**: N/A

**Comment by RalfJung** [comment]
Ah, here's the one in the rustc repo, revision fbdb10f9fabe47eb763cb4b52b5721740cc63783.
```
[{
	"resource": "/home/r/src/rust/rustc/compiler/rustc_middle/src/mir/interpret/allocation.rs",
	"owner": "rustc",
	"code": {
		"value": "type-mismatch",
		"target": {
			"$mid": 1,
			"external": "https://rust-analyzer.github.io/manual.html#type-mismatch",
			"path": "/manual.html",
			"scheme": "https",
			"authority": "rust-analyzer.github.io",
			"fragment": "type-mismatch"
		}
	},
	"severity": 8,
	"message": "expected &Allocation<Tag, Extra>, found Interned<Allocation<Tag, Extra>>",
	"source": "rust-analyzer",
	"startLineNumber": 76,
	"startColumn": 9,
	"endLineNumber": 76,
	"endColumn": 17
}]
```

**Comment by iDawer** [comment]
Minimized from miri:
```rust
fn main() {
    let x: &Option<()> = &Some(());
    let _: Option<()> = x.clone();
} //                    ^^^^^^^^^
```
rust-analyzer 46d7ee68f 2022-04-04 stable + rustc 1.62.0-nightly (69a5d2481 2022-04-27): `expected Option<()>, found &Option<()>`.
rust-analyzer 241b9f9ae 2022-04-28 nightly + rustc 1.62.0-nightly (69a5d2481 2022-04-27): good. Seems fixed?




> Ah, here's the one in the rustc repo, revision fbdb10f9fabe47eb763cb4b52b5721740cc63783, in [this line](https://github.com/rust-lang/rust/blob/dfb4194e3b01479c80f32586043264abd86fb1ab/compiler/rustc_middle/src/mir/interpret/allocation.rs#L76).

This one is #1109 or #10560

**Comment by flodiebold** [comment]
The first one was #11936.
